ferrocene/tools/traceability-matrix/README.md

```diff
@@ -0,0 +1,101 @@
+<!-- SPDX-License-Identifier: MIT OR Apache-2.0 -->
+<!-- SPDX-FileCopyrightText: The Ferrocene Developers -->
+
+# traceability-matrix
+
+The traceability matrix is part of our qualification material. You can find it at https://public-docs.ferrocene.dev/main/qualification/traceability-matrix.html.
+
+In the overview it states that it is:
+
+> A matrix that ties the specifics of the Ferrocene Language Specification to the test suite. This was written to ensure that all documented requirements are covered.
+
+While it says "written" it is (fortunately) generated semi-automatically.
+
+How does it work?
+
+On the one hand we have the Ferrocene language specification (FLS). Simply speaking the FLS is a list of statements, which are grouped in sections and subsections. Each section, subsection and statement has a unique id, which look something like `fls_wt81sbsecmu0` or `fls_3xvm61x0t251`.
+
+On the other hand we have tests. We need to find at least one test for each statement from the FLS. This test is then annotated with the id of the statement. This annotating is happening through a comment which is getting added to the end of the file. It is possible to annotate one test with multiple ids.
+
+For example `tests/ui/borrowck/borrowck-break-uninit-2.rs` got annotated with `fls_3xvm61x0t251` ([file at c80de1f](https://github.com/ferrocene/ferrocene/blob/c80de1fa7eecdfbe4579b13e2aaa93fb0586f9c6/tests/ui/borrowck/borrowck-break-uninit-2.rs)):
+
+```rust
+fn foo() -> isize {
+    let x: isize;
+
+    while 1 != 2  {
+        break;
+        x = 0;
+    }
+
+    println!("{}", x); //~ ERROR E0381
+
+    return 17;
+}
+
+fn main() { println!("{}", foo()); }
+
+//
+// ferrocene-annotations: fls_3xvm61x0t251
+// Initialization
+```
+
+It is also possible to annotate bulks of tests. This happens through `ferrocene-annotations` files, which basically contain the same comments. For example `tests/ui/borrowck/ferrocene-annotations` contains three ids:
+
+```
+// ferrocene-annotations: fls_a14slch83hzn
+// Borrowing
+//
+// ferrocene-annotations: fls_411up5z0b6n6
+// Lexical Elements
+//
+// ferrocene-annotations: fls_fgnllgz5k3e6
+// Lexical Elements, Separators, and Punctuation
+```
+
+This annotates all the tests in `tests/ui/borrowck/` with those three ids.
+
+When using bulk annotation, individual tests can still be annotated with additional ids, as is the case for `tests/ui/borrowck/borrowck-break-uninit-2.rs`. It is annotated with four ids. The three from `tests/ui/borrowck/ferrocene-annotations` and one from the file itself.
+
```

Comment:
note that the third annotation type is run-make tests (found in some makefiles, which are found in tests/run-make dir)

Reply:
annotations are written in a Makefile ([example](https://github.com/ferrocene/ferrocene/blob/87b1097519af7e72eb7fc91449f5c51dc9fbf462/tests/run-make/fpic/Makefile]), so-called [run-make tests](https://rustc-dev-guide.rust-lang.org/tests/compiletest.html#run-make-tests)